### ISSUE TITLE
Static typing with cimport introduces unnecessary dependencies

### ISSUE THREAD

**Issue by trouger** [cython/cython#1534]
Currently, when using Cython on existing python scripts, if we want to add static typing for a variable with a type from another module, we must add a line to cimport the type from that module. However, cimport has the side effect of importing and executing the module, even if we don't really need this. And this introduces additional dependencies between modules, which often requires much refactoring effort to break circular dependencies.

For example, when static typing a variable in module M1 with type T from module M2, the only run-time information that is required by M1 is the address of the PyTypeObject struct of T, which is used for type checking. And this address can be retrieved without executing the python code in M2, because the PyTypeObject struct is a static variable in the compiled binary. Unless we have python code in M1 that subclasses T or instantiates T, etc., we don't have to execute M2 during executing M1.

In my humble opinion, we need some new Cython statement that provides the semantic of (cimport minus import). Maybe we should name it "creference"? For now I just hack into Cython to make cimport act like a creference, and if I want the original behaviour of cimport, I just put an import statement besides a cimport statement.

**Comment by robertwb** [comment]
There is type initialization code that happens on cimport--we'd rather not introduce a state where types or other cimported values are partially initialized. This would also mean that `cdef T var` would be legal, but `cdef T var = T()` would not be. I think it's much better for a symbol to be all the way available if it's available at all. Also, the bar for introducing new Cython syntax and concepts is high.

**Comment by trouger** [comment]
@robertwb That's right, I understand introducing a partially available type is not the best choice, and my current solution is to raise a NameError if a "creferenced" type is being instantiated and so on (such a type is considered to only be available as a static type annotation, but not available in the python namespace).

Although my solution may not be the best choice, I sincerely suggest that the dependency problem is worth considering. Cython is a very good language and I use it in my project to gain large performance boost for one or two core modules. But if I want to use it more widely, maybe for an existing python framework that consists of tens of modules, I often find it extremely difficult, sometimes impossible, to break circular dependencies that are introduced through cimport. For example, we often have modules that have a dependency chain like M1->M2->M3->M4->M5, and objects of types from these modules are often passed between these modules, then static typing will introduce new dependencies, e.g., M4->M2, and M5->M1, ..., and then we have to either (1) refactor these modules heavily to utilize Cython, but with penalty in readability and extensibility, or (2) giving up static typing across modules.

I'd really appreciate it if Cython could provide a good solution for this problem.

**Comment by robertwb** [comment]
I've generally found that refactoring modules to eliminate circular dependencies often increases readability and structure, bug agree that it's non-trivial work. 

In this particular case (going on the scant details I have), I would suggest introducing a base class (almost an interface) in M1 which M5 then implements. 

As for a long-term solution, I think it'd be interesting to explore Go-style interfaces and fast calling without explicit cimports.
